docs/decisions/0020-upstream-downstream.rst

```diff
@@ -0,0 +1,373 @@
+        Mixed into CMS's XBLOCK_MIXINS, but not LMS's.
+        """
+
+        # Metadata related to upstream synchronization
+        upstream = String(
+            help=("""
+                The usage key of a block (generally within a content library)
+                which serves as a source of upstream updates for this block,
+                or None if there is no such upstream. Please note: It is valid
+                for this field to hold a usage key for an upstream block
+                that does not exist (or does not *yet* exist) on this instance,
+                particularly if this downstream block was imported from a
+                different instance.
+            """),
+            default=None, scope=Scope.settings, hidden=True, enforce_type=True
+        )
+        upstream_version = Integer(
+            help=("""
+                Record of the upstream block's version number at the time this
+                block was created from it. If upstream_version is smaller
+                than the upstream block's latest version, then the user will be
+                able to sync updates into this downstream block.
+            """),
+            default=None, scope=Scope.settings, hidden=True, enforce_type=True,
+        )
+        downstream_customized = List(
+            help=("""
+                Names of the fields which have values set on the upstream
+                block yet have been explicitly overridden on this downstream
+                block. Unless explicitly cleared by the user, these
+                customizations will persist even when updates are synced from
+                the upstream.
+            """),
+            default=[], scope=Scope.settings, hidden=True, enforce_type=True,
+        )
+
+        # Store upstream defaults for customizable fields.
+        upstream_display_name = String(...)
+        upstream_max_attempts = List(...)
+        ...  # We will probably want to pre-define several more of these.
+
+        def get_upstream_field_names(cls) -> dict[str, str]:
+            """
+            Mapping from each customizable field to field which stores its upstream default.
+            XBlocks outside of edx-platform can override this in order to set
+            up their own customizable fields.
+            """
+            return {
+                "display_name": "upstream_display_name",
+                "max_attempts": "upstream_max_attempts",
+            }
+
+        def save(self, *args, **kwargs):
+            """
+            Update `downstream_customized` when a customizable field is modified.
+            Uses `get_upstream_field_names` keys as the list of fields that are
+            customizable.
+            """
+            ...
+
+        @XBlock.handler
+        def sync_updates(self, request: Request, suffix=None) -> Response:
+            """
+            POST handler which applies updates from the upstream block.
+            No request/response data.
```

Comment:
I don't think this should be an `@XBlock.handler`, unless we _want_ to support the case where an XBlock completely overrides what it means to sync updates for that type of block. I'd rather move this to the runtime or the `content_libraries` app or a `content_sync` app and enforce that the syncing behavior of all blocks is always consistent.

We can call a `post_sync_handler()` method on each block afterward if you want a slot for customization though.

Reply:
I'm in fuzzy agreement with @bradenmacdonald on this, but I don't have a great sense for where it _should_ go. Was it in the handler here because we originally thought individual XBlocks would be required to override this to get the right behavior?

Reply:
@ormsbee @bradenmacdonald 

> Was it in the handler here because we originally thought individual XBlocks would be required to override this to get the right behavior?

No, that was not the intention-- I feel pretty strongly that individual XBlock types should **not** get to customize this behavior 😄 I put this here because it was a convenient way of adding the sync capability to every CMS block without adding it LMS block.

> the content_libraries app or a content_sync app 

Right, so I'm trying to strike a balance between (a) having an abstraction layer between syncing and content_libraries, while (b) keeping the whole upstream-downstream idea lightweight since it's so nascent. Making a new content_sync app seemed like it would be overengineering.

> I don't have a great sense for where it *should* go.

Here's what I'm thinking now: we define a CMS-only HTTP API for syncing in cms/djangoapps/contentstore.
* In the case of LC->ModuleStore syncing of components (Sumac), the HTTP API calls out to a new method, `CachingDescriptorSystem.sync_from_upstream`.
* In the case of LC->LC syncing of components (future), the HTTP API would call out to `LearningCoreRuntime.sync_from_upstream`. Generally speaking, syncing from upstream component blocks to downstream component blocks would be done by calling the abstract `Runtime.sync_from_upstream` method on the *downstream* runtime.

Thoughts?

Reply:
Makes sense to me.

Reply:
I think that's fine.

Reply:
Added to the ADR